operator/pkg/controllers/etcd/etcd.go

```diff
@@ -20,181 +20,77 @@ import (
 
-func New(kubeclient client.Client) *Controller {
-	return &Controller{kubeClient: kubeclient, secretsProvider: secrets.New(kubeclient)}
-}
+type reconciler func(ctx context.Context, controlPlane *v1alpha1.ControlPlane) (err error)
 
-func (c *Controller) Reconcile(ctx context.Context, controlPlane *v1alpha1.ControlPlane) (err error) {
-	// generate etcd service object
-	if err := c.createService(ctx, controlPlane); err != nil {
-		return err
-	}
-	// Create ETCD certs and keys, store them as secret in the management server
-	if err := c.createSecrets(ctx, controlPlane); err != nil {
-		return err
-	}
-	// create etcd stateful set
-	if err := c.createStatefulset(ctx, controlPlane); err != nil {
-		return err
-	}
-	return nil
+func New(kubeclient *kubeprovider.Client) *Controller {
+	return &Controller{kubeClient: kubeclient, certificates: common.New(kubeclient)}
 }
 
-func (c *Controller) createService(ctx context.Context, controlPlane *v1alpha1.ControlPlane) error {
-	svc := &v1.Service{
-		ObjectMeta: metav1.ObjectMeta{
-			Name:      etcdServiceNameFor(controlPlane.ClusterName()),
-			Namespace: controlPlane.NamespaceName(),
-			Labels:    etcdLabelFor(controlPlane.ClusterName()),
-			OwnerReferences: []metav1.OwnerReference{{
-				APIVersion: controlPlane.APIVersion,
-				Name:       controlPlane.Name,
-				Kind:       controlPlane.Kind,
-				UID:        controlPlane.UID,
-			}},
-		},
-	}
-	result, err := controllerutil.CreateOrPatch(ctx, c.kubeClient, svc, func() error {
-		// We can't update the Spec field completely as the existing svc object
-		// has some defaults set by API server like `Spec.Type: ClusterIP`. If
-		// we update Spec field, we will need to set these defaults as well
-		// because CreateOrPatch does a reflect.DeepEqual for the existing spec
-		// and with our change, calls Patch if they are not equal.
-		svc.Spec.Selector = etcdLabelFor(controlPlane.ClusterName())
-		svc.Spec.Ports = []v1.ServicePort{{
-			Port:       2380,
-			Name:       etcdServerPortNameFor(controlPlane.ClusterName()),
-			TargetPort: intstr.IntOrString{Type: intstr.Int, IntVal: 2380},
-			Protocol:   "TCP",
-		}, {
-			Port:       2379,
-			Name:       etcdClientPortNameFor(controlPlane.ClusterName()),
-			TargetPort: intstr.IntOrString{Type: intstr.Int, IntVal: 2379},
-			Protocol:   "TCP",
-		}}
-		svc.Spec.ClusterIP = "None"
-		return nil
-	})
-	if err != nil {
-		return fmt.Errorf("creating service %s/%s, %w", svc.Namespace, svc.Name, err)
-	}
-	if result != controllerutil.OperationResultNone {
-		zap.S().Infof("[%s] service %s %s", controlPlane.ClusterName(), svc.Name, result)
+func (c *Controller) Reconcile(ctx context.Context, controlPlane *v1alpha1.ControlPlane) (err error) {
+	for _, reconcile := range []reconciler{
+		c.reconcileService,
+		c.reconcileSecrets,
+		c.reconcileStatefulSet,
+	} {
+		if err := reconcile(ctx, controlPlane); err != nil {
+			return err
+		}
 	}
+	zap.S().Infof("[%v] etcd reconciled", controlPlane.ClusterName())
```

Comment:
These types of logs are gonna get a bit noisy. Consider only logging when you take explicit action.

Reply:
Will revisit this later.